File: download_images.py
#!/usr/bin/env python3
import os
import json
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from urllib.request import urlopen, urlretrieve, Request

TOTAL = 1025
CONCURRENCY = 8
OUTDIR = os.path.join(os.path.dirname(__file__), 'images')
if not os.path.exists(OUTDIR):
    os.makedirs(OUTDIR)

def fetch_json(url):
    req = Request(url, headers={'User-Agent':'python-urllib/3'})
    with urlopen(req, timeout=30) as resp:
        return json.load(resp)

def download_image(url, dest):
    try:
        req = Request(url, headers={'User-Agent':'python-urllib/3'})
        urlretrieve(url, dest)
        return True
    except Exception as e:
        return False

def process(id):
    fname = os.path.join(OUTDIR, f"{id:03}.png")
    if os.path.exists(fname):
        return (id, 'skipped')
    try:
        data = fetch_json(f'https://pokeapi.co/api/v2/pokemon/{id}')
        url = None
        url = data.get('sprites', {}).get('other', {}).get('official-artwork', {}).get('front_default')
        if not url:
            url = data.get('sprites', {}).get('front_default')
        if not url:
            return (id, 'no-image')
        ok = download_image(url, fname)
        return (id, 'ok' if ok else 'failed')
    except Exception as e:
        return (id, 'error')

def main():
    print(f'Starting download of {TOTAL} images with concurrency={CONCURRENCY}')
    successes = 0
    fails = 0
    skipped = 0
    with ThreadPoolExecutor(max_workers=CONCURRENCY) as ex:
        futures = {ex.submit(process, i): i for i in range(1, TOTAL+1)}
        for fut in as_completed(futures):
            id = futures[fut]
            try:
                id, status = fut.result()
            except Exception as e:
                print(f'id={id} -> exception')
                fails += 1
                continue
            if status == 'ok':
                successes += 1
            elif status == 'skipped':
                skipped += 1
            else:
                fails += 1
            print(f'id={id:03} -> {status} (done={successes} fail={fails} skip={skipped})')
    print('Finished. success=', successes, 'fail=', fails, 'skipped=', skipped)

if __name__ == '__main__':
    main()

File: codigos/download_images.js
// Script para baixar as imagens oficiais (official-artwork) de 1..1025
// Uso: `node download_images.js`

const fs = require('fs');
const path = require('path');
const https = require('https');

const TOTAL = 1025;
const CONCURRENCY = 8;
const OUTDIR = path.join(__dirname, 'images');

if(!fs.existsSync(OUTDIR)) fs.mkdirSync(OUTDIR);

async function fetchJson(url){
  return fetch(url).then(r=>{
    if(!r.ok) throw new Error(`${r.status}`);
    return r.json();
  });
}

function downloadUrlToFile(url, dest){
  return new Promise((resolve,reject)=>{
    const file = fs.createWriteStream(dest);
    https.get(url, res =>{
      if(res.statusCode !== 200){
        file.close(); fs.unlinkSync(dest); return reject(new Error('status '+res.statusCode));
      }
      res.pipe(file);
      file.on('finish', ()=> file.close(resolve));
    }).on('error', err =>{
      try{ file.close(); fs.unlinkSync(dest);}catch(e){}
      reject(err);
    });
  });
}

async function processId(id){
  const out = path.join(OUTDIR, `${String(id).padStart(3,'0')}.png`);
  if(fs.existsSync(out)) return {id, skipped:true};
  try{
    const data = await fetchJson(`https://pokeapi.co/api/v2/pokemon/${id}`);
    const url = data.sprites?.other?.['official-artwork']?.front_default || data.sprites?.front_default;
    if(!url) return {id, ok:false};
    await downloadUrlToFile(url, out);
    return {id, ok:true};
  }catch(err){
    return {id, ok:false, err: String(err)};
  }
}

async function run(){
  console.log('Iniciando download de imagens para', TOTAL, 'pokémons');
  let i=1;
  let success=0, fail=0, skipped=0;
  const queue = [];
  while(i<=TOTAL){
    while(queue.length < CONCURRENCY && i<=TOTAL){
      const id = i++;
      const p = processId(id).then(res=>{
        queue.splice(queue.indexOf(p),1);
        if(res.skipped) skipped++;
        else if(res.ok) success++; else fail++;
        console.log(`id=${res.id} ok=${res.ok} skipped=${!!res.skipped}` + (res.err? ' err='+res.err:''));
      });
      queue.push(p);
    }
    await Promise.race(queue).catch(()=>{});
  }
  await Promise.allSettled(queue);
  console.log('Concluído. success=', success, 'fail=', fail, 'skipped=', skipped);
}

run().catch(err=>{ console.error(err); process.exit(1); });
